### PR TITLE
backupccl: distribute RESTORE work using DistSQL

### PR DESCRIPTION
**Pull request by pbardea**
This commit replaces restore's old method of coordinating work on a
single node to instead use DistSQL. It creates a 2 stage DistSQL flow.
The first stage assigns chunks to SplitAndScatter processors which
forward spans that they scattered to the second stage made up of the
RestoreData processors which ingest the data. The SplitAndScatter
processors attempt to send the work to the RestoreData colocated with
the leaseholder of the span after the scattering.

Release note: None.

### CODE REVIEW

**Comment by pbardea** [comment]
This should be ready for a review, but shouldn't be merged until #51607 is merged and the scatter processor can actually find the destination of the span it imported.

**Review by dt — APPROVED** [review]

File: pkg/ccl/backupccl/restore_job.go [hunk 1]
```diff
@@ -598,61 +445,57 @@ func restore(
...
 		syncutil.Mutex
-		res               RowCount
-		requestsCompleted []bool
-		highWaterMark     int
+		highWaterMark int
 	}{
 		highWaterMark: -1,
 	}
 
+	res := RowCount{}
```
[the remark this answers is not in the document]
Comment (pbardea) [reply pkg/ccl/backupccl/restore_job.go]: Done - also moved back `requestsCompleted` for similar reasons. I also am now asserting that the dataspan we're marking as done is the one that has the same progress idx.

**Review by yuzefovich — COMMENTED** [review]
LGTM

> Reviewed 3 of 4 files at r1, 3 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51959#-:-MDLIYhB--AwaN4PTzyl:bb6ywmx)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt, @pbardea, and @yuzefovich)

___
*[pkg/ccl/backupccl/restore_processor_planning.go, line 17 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51959#-MDL7Swe8HaRbQhTCniS:-MDL7Swe8HaRbQhTCniT:b-c8kygd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63453445c747a5d9952e50e2fd2b37a28b3c74b/pkg/ccl/backupccl/restore_processor_planning.go#L17)):*
> ```Go
> 
> 	"github.com/cockroachdb/cockroach/pkg/kv"
> 	roachpb "github.com/cockroachdb/cockroach/pkg/roachpb"
> ```

super nit: are these `roachpb` and `hlc` aliases necessary?
___
*[pkg/ccl/backupccl/restore_processor_planning.go, line 71 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51959#-MDLHADc8o5Lt_WRSzsC:-MDLHADc8o5Lt_WRSzsD:b-w01eyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63453445c747a5d9952e50e2fd2b37a28b3c74b/pkg/ccl/backupccl/restore_processor_planning.go#L71)):*
> ```Go
> 
> 	if len(splitAndScatterSpecs) == 0 {
> 		close(progCh)
> ```

nit: we probably can defer this close. I'd also mention in the comment to the method that `progCh` is closed.

**Review by pbardea — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51959)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @yuzefovich)

___
*[pkg/ccl/backupccl/restore_processor_planning.go, line 17 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51959#-MDL7Swe8HaRbQhTCniS:-MDLKWkP4g3nB-SA03qy:bmuaccw) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63453445c747a5d9952e50e2fd2b37a28b3c74b/pkg/ccl/backupccl/restore_processor_planning.go#L17)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

super nit: are these `roachpb` and `hlc` aliases necessary?
</blockquote></details>

Nope, those were somehow auto-generated. Thanks for catching!
___
*[pkg/ccl/backupccl/restore_processor_planning.go, line 71 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/51959#-MDLHADc8o5Lt_WRSzsC:-MDLK_FM0LbWnGWaMGZk:bx4nh62) ([raw file](https://github.com/cockroachdb/cockroach/blob/c63453445c747a5d9952e50e2fd2b37a28b3c74b/pkg/ccl/backupccl/restore_processor_planning.go#L71)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: we probably can defer this close. I'd also mention in the comment to the method that `progCh` is closed.
</blockquote></details>

Done and done.

**Comment by pbardea** [comment]
TFTRs!
bors r=dt,yuzefovich

**Pull request merged**